spv/sync.go

```diff
@@ -0,0 +1,1430 @@
+		s.notifications.FetchHeadersProgress(int32(lastHeader.Height), lastHeader.Timestamp.Unix())
+	}
+}
+
+func (s *Syncer) fetchHeadersFinished() {
+	if s.notifications != nil && s.notifications.FetchHeadersFinished != nil {
+		s.notifications.FetchHeadersFinished()
+	}
+}
+func (s *Syncer) discoverAddressesStart(walletID int) {
+	if s.notifications != nil && s.notifications.DiscoverAddressesStarted != nil {
+		s.notifications.DiscoverAddressesStarted(walletID)
+	}
+}
+
+func (s *Syncer) discoverAddressesFinished(walletID int) {
+	if s.notifications != nil && s.notifications.DiscoverAddressesFinished != nil {
+		s.notifications.DiscoverAddressesFinished(walletID)
+	}
+}
+
+func (s *Syncer) rescanStart(walletID int) {
+	if s.notifications != nil && s.notifications.RescanStarted != nil {
+		s.notifications.RescanStarted(walletID)
+	}
+}
+
+func (s *Syncer) rescanProgress(walletID int, rescannedThrough int32) {
+	if s.notifications != nil && s.notifications.RescanProgress != nil {
+		s.notifications.RescanProgress(walletID, rescannedThrough)
+	}
+}
+
+func (s *Syncer) rescanFinished(walletID int) {
+	if s.notifications != nil && s.notifications.RescanFinished != nil {
+		s.notifications.RescanFinished(walletID)
+	}
+}
+
+func (s *Syncer) mempoolTxs(walletID int, txs []*wire.MsgTx) {
+	if s.notifications != nil && s.notifications.MempoolTxs != nil {
+		s.notifications.MempoolTxs(walletID, txs)
+	}
+}
+
+func (s *Syncer) tipChanged(tip *wire.BlockHeader, reorgDepth int32, matchingTxs map[chainhash.Hash][]*wire.MsgTx) {
+	if s.notifications != nil && s.notifications.TipChanged != nil {
+		var txs []*wire.MsgTx
+		for _, matching := range matchingTxs {
+			txs = append(txs, matching...)
+		}
+		s.notifications.TipChanged(tip, reorgDepth, txs)
+	}
+}
+
+func (s *Syncer) lowestChainTip(ctx context.Context) (chainhash.Hash, int32, *wallet.Wallet) {
+	var lowestTip int32 = -1
+	var lowestTipHash chainhash.Hash
+	var lowestTipWallet *wallet.Wallet
+	for _, w := range s.wallets {
+		if hash, height := w.MainChainTip(ctx); height < lowestTip || lowestTip == -1 {
+			lowestTip = height
+			lowestTipHash = hash
+			lowestTipWallet = w
+		}
+	}
+
+	return lowestTipHash, lowestTip, lowestTipWallet
+}
+
+func (s *Syncer) highestChainTip(ctx context.Context) (chainhash.Hash, int32, *wallet.Wallet) {
+	var highestTip int32 = -1
+	var highestTipHash chainhash.Hash
+	var highestTipWallet *wallet.Wallet
+	for _, w := range s.wallets {
+		if hash, height := w.MainChainTip(ctx); height > highestTip || highestTip == -1 {
+			highestTip = height
+			highestTipHash = hash
+			highestTipWallet = w
+		}
+	}
+
+	return highestTipHash, highestTip, highestTipWallet
+}
+
+// Run synchronizes the wallet, returning when synchronization fails or the
+// context is cancelled.
+func (s *Syncer) Run(ctx context.Context) error {
+	tipHash, tipHeight, lowestChainWallet := s.lowestChainTip(ctx)
+	rescanPoint, err := lowestChainWallet.RescanPoint(ctx)
```

Comment:
The comment below in lines 311-312 explain why we're fetching this `rescanPoint` - just to help us determine the last point that txs were synced up to. It is theoretically possible that another wallet with more block headers has a lower rescan point. I think `s.lowestRescanPoint` should be used here.
```suggestion
	lowestRescanPoint, err := s.lowestRescanPoint(ctx)
```